src/backend/distributed/executor/adaptive_executor.c

```diff
@@ -738,17 +738,31 @@ RunLocalExecution(CitusScanState *scanState, DistributedExecution *execution)
 static void
 AdjustDistributedExecutionAfterLocalExecution(DistributedExecution *execution)
 {
-	/*
-	 * Local execution already stores the  tuples for returning, so we should not
-	 * store again.
-	 */
-	execution->hasReturning = false;
-
 	/* we only need to execute the remote tasks */
 	execution->tasksToExecute = execution->remoteTaskList;
 
 	execution->totalTaskCount = list_length(execution->remoteTaskList);
 	execution->unfinishedTaskCount = list_length(execution->remoteTaskList);
+
+	/*
+	 * For reference tables, local execution already stores the  tuples for
+	 * returning, so we should not store again.
+	 *
+	 * We typically expect to have modifications to reference tables when the
+	 * unfinishedTaskCount is one. However, we still explicitly check for
+	 * reference tables because there are some edge cases for distributed
+	 * tables with 2 shards.
+	 */
+	if (execution->hasReturning && execution->unfinishedTaskCount == 1)
+	{
+		Task *initialTask = (Task *) linitial(execution->tasksToExecute);
+
+		if (initialTask->anchorShardId != INVALID_SHARD_ID &&
+			ReferenceTableShardId(initialTask->anchorShardId))
```

Comment:
This fix seems messy. Could we keep track of which tasks were split up? e.g. set a flag on the Task in 	`ExtractLocalAndRemoteTasks`?

Reply:
An alternative is that we create the ShardCommandExecution states before the local execution and keep track in there.

Reply:
> This fix seems messy. Could we keep track of which tasks were split up? e.g. set a flag on the Task in ExtractLocalAndRemoteTasks?

I considered that as well, which makes sense. But isn't it already equivalent for `ReferenceTableShardId(initialTask->anchorShardId) && taskType == ROUTER_TASK`? Since we could have both local and remote tasks only for modifications, I skipped checking  `taskType == ROUTER_TASK`. I think `hasReturning` check already also ensures that this is a modification.

(I'm trying to avoid adding things to `Task` which is only relevant for a very specific case, but if you think that's OK, I can do that)

Reply:
I think making the notion of "this task is part local, part remote" explicit would be helpful in making code more understandable.

Reply:
Added the explicit field, and moved the decision to `ShardCommandExecution `, which makes more sense?